### PR TITLE
GUI multi-document support 

### LINKED ISSUE #77: Allow bulk conversion to safe PDFs
A few different people have requested this feature: the ability to convert documents to safe PDFs in bulk.

This shouldn't be too difficult to implement, but I think the biggest issue is how the user interface should work. One idea is to allow the user to browse for a folder, and then try to automatically convert all documents it can in that folder to save versions (appending `-safe.pdf` to the end). It probably makes sense to convert them one at a time, and also display a report of all unsupported files in the folder that couldn't get converted.

What happens if there's a file called `test.pdf` and another called `test-safe.pdf` in the same folder though? What would the safe version of `test.pdf` get named? It could maybe detect this edge case, and tell you that you must not have any files that end in `-safe.pdf` in the folder to convert.

### LINKED ISSUE #205: 0.3.1 release: parallel document conversion fails
If we open two Dangerzone windows via the tray menu's "new window" both windows get closed as soon as the first document finishes the conversion.

Release: 0.3.1

### Terminal log

From taking a look at the log, it seems it forcefully exits the whole application, leaving the conversion threat without being properly terminated or waited for.

```
$ poetry run dev_scripts/dangerzone

(python3:78763): Gdk-CRITICAL **: 16:08:27.352: gdk_atom_intern: assertion 'atom_name != NULL' failed

(python3:78763): Gdk-CRITICAL **: 16:08:27.353: gdk_atom_intern: assertion 'atom_name != NULL' failed
> /usr/bin/podman run --network none -v /home/user/dangerzone/test_docs/sample.pptx:/tmp/input_file -v /home/user/.config/dangerzone/tmp/tmpojlc23em/pixels:/dangerzone dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py document-to-pixels
0% Converting to PDF using LibreOffice
3% Separating document into pages
5% Converting page 1/3 to pixels
20% Converting page 2/3 to pixels
35% Converting page 3/3 to pixels
50% Converted document to pixels
> /usr/bin/podman run --network none -v /home/user/.config/dangerzone/tmp/tmpojlc23em/pixels:/dangerzone -v /home/user/.config/dangerzone/tmp/tmpojlc23em/safe:/safezone -e OCR=1 -e OCR_LANGUAGE=eng dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py pixels-to-pdf
50% Converting page 1/3 from pixels to searchable PDF
> /usr/bin/podman run --network none -v /home/user/dangerzone/test_docs/sample.odp:/tmp/input_file -v /home/user/.config/dangerzone/tmp/tmp_6uk9hvq/pixels:/dangerzone dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py document-to-pixels
0% Converting to PDF using LibreOffice
65% Converting page 2/3 from pixels to searchable PDF
80% Converting page 3/3 from pixels to searchable PDF
3% Separating document into pages
5% Converting page 1/3 to pixels
20% Converting page 2/3 to pixels
35% Converting page 3/3 to pixels
50% Converted document to pixels
> /usr/bin/podman run --network none -v /home/user/.config/dangerzone/tmp/tmp_6uk9hvq/pixels:/dangerzone -v /home/user/.config/dangerzone/tmp/tmp_6uk9hvq/safe:/safezone -e OCR=1 -e OCR_LANGUAGE=eng dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py pixels-to-pdf
50% Converting page 1/3 from pixels to searchable PDF
65% Converting page 2/3 from pixels to searchable PDF
95% Merging 3 pages into a single PDF
80% Converting page 3/3 from pixels to searchable PDF
97% Compressing PDF
100% Safe PDF created
> evince /home/user/dangerzone/test_docs/sample-safe.pdf
QThread: Destroyed while thread is still running

(evince:79347): Gdk-CRITICAL **: 16:08:49.822: gdk_atom_intern: assertion 'atom_name != NULL' failed

(evince:79347): Gdk-CRITICAL **: 16:08:49.822: gdk_atom_intern: assertion 'atom_name != NULL' failed
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by deeplow**
Fixes #77 and UX inspired by #117. Depends on #216. Fixes #205.

dz-run-gui2

- adds bulk conversion to the GUI
- removes "new window" logic (previous way of doing parallel conversions)
- updates settings to make sense in a multi-document scenario

### CODE REVIEW

**Comment by deeplow** [comment]
Rebased this on top of https://github.com/freedomofpress/dangerzone/pull/216

File: dangerzone/gui/__init__.py [hunk 1]
```diff
@@ -122,3 +123,23 @@ def application_activated() -> None:
 
 def setup_logging() -> None:
     logging.basicConfig(level=logging.DEBUG, format="[%(levelname)s] %(message)s")
+
+def override_arg_parser() -> None:
```
Comment (apyrgio) [inline dangerzone/gui/__init__.py:109]: This should change once we merge #216.

File: dangerzone/document.py [hunk 2]
```diff
@@ -103,6 +103,31 @@ def default_output_filename(self) -> str:
 
     def announce_id(self) -> None:
         log.info(f"Assigning ID '{self.id}' to doc '{self.input_filename}'")
+    def set_output_dir(self, path: str) -> None:
+        if not self._output_filename:
+            self.set_default_output_filename()
```
Comment (apyrgio) [inline dangerzone/document.py:108]: Most probably we shouldn't have any need for this function call, once we merge #216.

File: dangerzone/document.py [hunk 3]
```diff
@@ -103,6 +103,31 @@ def default_output_filename(self) -> str:
 
     def announce_id(self) -> None:
         log.info(f"Assigning ID '{self.id}' to doc '{self.input_filename}'")
+    def set_output_dir(self, path: str) -> None:
```
Comment (apyrgio) [inline dangerzone/document.py:106]: Should we check at some point if the output filenames from different dirs end up being the same name. This can happen if we have the same filename in different dirs.

File: dangerzone/document.py [hunk 4]
```diff
@@ -103,6 +103,31 @@ def default_output_filename(self) -> str:
...
+        if not self._output_filename:
+            self.set_default_output_filename()
+
+        # keep the same name
+        old_filename = os.path.basename(self.output_filename)
+
+        new_path = os.path.abspath(path)
+        if not os.path.exists(new_path):
+            raise errors.NonExistantOutputDirException()
```
Comment (apyrgio) [inline dangerzone/document.py:115]: Not that important, but we start to have various errors for different aspects of a document (input file, output file, output dir), so I think it's best to start namespacing them:

* `InputFile...Exception`
* `OutputFile...Exception`
* `OutputDir...Exception`

Probably not actionable now, but something to have in mind.

**Review by apyrgio — COMMENTED** [review]
I did a first pass of the code. I don't see something sticking out sorely, which is good. I'm pretty sure I'm not that thorough on the Qt side of things though. I hope that by experimenting a bit with it live, I'll understand more.

At some point we should polish a bit the commit messages, but that's to be expected since you created the PR before merging the previous one you've sent.

File: dangerzone/gui/__init__.py [hunk 5]
```diff
@@ -49,9 +49,15 @@ def monkeypatch_event(arg__1: QtCore.QEvent) -> bool:
...
 
 @click.command()
-@click.argument("filename", required=False, callback=args.validate_input_filename)
+@click.argument(
+    "filenames",
+    required=False,
+    nargs=-1,
+    type=click.UNPROCESSED,
+    callback=args.validate_input_filenames,
+)
```
[the remark this answers is not in the document]
Comment (deeplow) [reply dangerzone/gui/__init__.py:57]: At some point I tried doing that by moving the click.arguments to `dangerzone/__init__.py` but then running the tests got quite complicated -- the `DANGERZONE_MODE` env var is not set when we run the `CliRunner` and so it would just use the default one.

Or something along these lines. Can't remember exactly.

File: dangerzone/gui/main_window.py [hunk 6]
```diff
@@ -232,14 +232,14 @@ def __init__(self, dangerzone: DangerzoneGui, document: Document) -> None:
 
         # Layout
         layout = QtWidgets.QVBoxLayout()
-        layout.addWidget(self.doc_selection_widget, stretch=1)
         layout.addWidget(self.settings_widget, stretch=1)
         layout.addWidget(self.convert_widget, stretch=1)
+        layout.addWidget(self.doc_selection_widget, stretch=1)
```
[the remark this answers is not in the document]
Comment (deeplow) [reply dangerzone/gui/main_window.py:237]: I think so. This changes the order in which the widgets are displayed. But change here ended up not affecting the final result, so we could undo it.

File: dangerzone/document.py (hunk 2 above)
Comment (deeplow) [reply dangerzone/document.py:108]: I think we still will need. On the UI there is a button to change the output directory of all the files. And that should call this one.

File: dangerzone/document.py (hunk 3 above)
Comment (deeplow) [reply dangerzone/document.py:106]: I don't think we need that at this point because the user should only be able to select from one directory. At least I am assuming that.

The one case that might affect us is if we have `file.pdf` and `file.gif`. Both will converge to the same one. This actually happens on the sample test files. So we should catch this.

File: dangerzone/gui/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (apyrgio) [reply dangerzone/gui/__init__.py:57]: Got it. It's not worth the extra hassle, since it's trivial.

File: dangerzone/document.py (hunk 2 above)
Comment (apyrgio) [reply dangerzone/document.py:108]: Ah, I see. In that case, we need a common function for this, since we also have the `default_output_filename` property that does a slightly different thing.

File: dangerzone/document.py (hunk 3 above)
Comment (apyrgio) [reply dangerzone/document.py:106]: > I don't think we need that at this point because the user should only be able to select from one directory. At least I am assuming that.

Oh, I hadn't realized that, you're right. That is, for the file browser, at least. You can always provide files in different directories through the CLI. And of course the filetype case you mentioned also holds.

**Comment by apyrgio** [comment]
I Just took a look at the fixes, and they look good to me. I have a few more things though, before we're good to merge:

1. Suggest a fix for this: https://github.com/freedomofpress/dangerzone/pull/247#discussion_r1020244410
2. Fix the duplicate "Assigning ID ..." message that we see in the logs.
3. Check the `QPixmap::scaled: Pixmap is a null pixmap` error that I see in my logs.
4. Debug why I don't see your icons in my development environment. Probably has to do with my `share/` folder.
5. Test it on Windows.

On a separate note, I find that `tesseract` times out consistently in my laptop when I pass multiple documents. Handling timeouts will require some thinking, but let's leave it for a subsequent PR.

**Comment by deeplow** [comment]
> Suggest a fix for this: GUI multi-document support  #247 (comment)
> Fix the duplicate "Assigning ID ..." message that we see in the logs.

I agree with the fixes.

> Check the `QPixmap::scaled: Pixmap is a null pixmap` error that I see in my logs.
> Debug why I don't see your icons in my development environment. Probably has to do with my share/ folder.

I am guessing these two are related. Let me know what you find. It probably has something to do with the `get_resource_path(filename)` and the `share` dir as you mention.

> Test it on Windows.

Started doing this now and I have spotted an issue already. Added a FIXUP for that https://github.com/freedomofpress/dangerzone/pull/247/commits/e2d87b670c338e6dbeff3bc1913da8e555863471

**Comment by deeplow** [comment]
> > Test it on Windows.
> 
> Started doing this now and I have spotted an issue already. Added a FIXUP for that [e2d87b6](https://github.com/freedomofpress/dangerzone/commit/e2d87b670c338e6dbeff3bc1913da8e555863471)

The Windows GUI bug should be fixed in https://github.com/freedomofpress/dangerzone/pull/247/commits/f7a36be0562997c2da3aff5ad41fda757be1aa27.

The only other Windows issue is that it's opening output file in the file explorer. When you have more than one file, this gets quite annoying.

**Comment by deeplow** [comment]
> The only other Windows issue is that it's opening output file in the file explorer. When you have more than one file, this gets quite annoying.

I am now experimenting with opening the PDF reader on Windows. If we could get this working, we'd be closer to behavior parity on windows and this issue would be solved.

**Comment by apyrgio** [comment]
I just saw your changes on the two Windows issues. I tested them out both on Windows and Linux, and I think we're fine. Only thing left is to squash the fixups, tidy up the commit messages, and we're good to merge!

**Comment by deeplow** [comment]
I will do that now.

**Comment by deeplow** [comment]
Rebased, squashed FIXUP commits, fixed some typos. The only thing missing is the CI giving the green light and merging.
